src/terraform-provider-signalform/signalform/integration.go

```diff
@@ -0,0 +1,116 @@
+package signalform
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"strings"
+)
+
+const (
+	INTEGRATION_API_URL = "https://api.signalfx.com/v2/integration"
+)
+
+func integrationResource() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of the integration",
+			},
+			"enabled": &schema.Schema{
+				Type:        schema.TypeBool,
+				Required:    true,
+				Description: "Whether the integration is enabled or not",
+			},
+			"type": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "Type of the integration",
+				ValidateFunc: validateIntegrationType,
+			},
+			"api_key": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "PagerDuty API key",
+				Sensitive:   true,
+			},
+			"webhook_url": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Slack Incoming Webhook URL",
+				Sensitive:   true,
+			},
+		},
+
+		Create: integrationCreate,
+		Read:   integrationRead,
+		Update: integrationUpdate,
+		Delete: integrationDelete,
+	}
+}
+
+func validateIntegrationType(v interface{}, k string) (we []string, errors []error) {
```

Comment:
Is there a reason, we are not validating the presence of `api_key` for PagerDuty and `webhook_url` for slack ? The check here is only done for integration type but I think it should also perform other necessary checks.

Reply:
@sjung-stripe sounds good. I think we can still ensure some validation in the schema itself with `ConflictsWith`. This is not really a blocker, but would be nice to have one level of protection here itself. 

Also, another issue is if we are to make this extendible to other integration, this schema might become a problem. We might have to end up modelling nested schema here for different integration with all the optional fields out there.

Ref (`ConflictsWith`): https://github.com/Yelp/terraform-provider-signalform/blob/master/src/terraform-provider-signalform/signalform/detector.go#L75

Reply:
> Also, another issue is if we are to make this extendible to other integration, this schema might become a problem. We might have to end up modelling nested schema here for different integration with all the optional fields out there.

I wanted to do something like this, but wasn't able to represent it in the schema. Do you have something concrete in mind?

re conflictswith - sure, I can add that.

Reply:
Yes we could do something like:

```go
"slack": &schema.Schema{
    Type:        schema.TypeSet,
    Optional:    true,
    Description: "slack integration",
    ConflictsWith: (all other named integration here)
    Elem: ... (whatever keys you want),
"pagerduty": .....(same as above)
```

In this case each integration now becomes a key. It will be optional, conflictWith other integrations ensuring there is only one integration type is defined. Further, you can now make the attributes such as `webhook_url` or `api_key` required within the integrations

Reply:
The problem with that approach is that it's using `schema.TypeSet`, which means this is now legal:

```
resource "signalform_integration" "slack" {
  name = "foo"
  enabled = true
}
```

ie you can specify zero copies of everything. You can't fix this with `ValidateFunc` either, because you'd have to check each of the keys (`slack`, `pagerduty`, etc) and make sure at least one was nonzero.

Reply:
I don't understand what you meant by this ? I was thinking of including "slack" like this (axis_right) https://github.com/Yelp/terraform-provider-signalform/blob/master/src/terraform-provider-signalform/signalform/time_chart.go#L115 not as a resource.

You can define the elemTypes you need now  and make them required. Further, make slack conflictWith the other integration types you define.

Reply:
Yes, the example you indicated has the same problem I'm describing. It's using a TypeSet which means it can be specified zero or more times. All of the following would be legal:

```
resource "signalform_integration "slack" {
  name = "foo"
  enabled = true
  slack {
    webhook_url = "foo"
  }
}
```

```
resource "signalform_integration "slack" {
  name = "foo"
  enabled = true
}
```

```
resource "signalform_integration "slack" {
  name = "foo"
  enabled = true
  slack {
    webhook_url = "foo"
  }
  slack {
    webhook_url = "bar"
  }
}
```

This is why I didn't use typed keys to begin with - because the only way to do so is to specify them on an `Elem`, which requires using a map, list, or set.

Reply:
Aah! understood what you meant now, Sorry about that. 

The project structure more or less follows this structure elsewhere (the one I mentioned above of using TypeSet). I feel that way, the structure is more appropriate than it currently is, considering the number of integrations out there and most of the attributes vary.

[signalform-tools](https://github.com/Yelp/signalform-tools) is another repository we maintain for validation and we can definitely bake in the functionality there.

This is not really a blocker, feel free to disregard.